registry/remote/repository.go

```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		} else {
+			// GET without server `Docker-Content-Digest` header forces the
+			// expensive calculation
+			if calculatedDigest, err = calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes); err != nil {
+				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+			}
+			contentDigest = calculatedDigest
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
-
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+	} else {
+		contentDigest = serverHeaderDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if len(refDigest) > 0 && len(contentDigest) > 0 && refDigest != contentDigest {
+		return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
 	}
 
+	// 6. Finally, if we made it this far, then all is good; return.
 	return ocispec.Descriptor{
 		MediaType: mediaType,
 		Digest:    contentDigest,
-		Size:      size,
+		Size:      resp.ContentLength,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response, maxMetadataBytes int64) (digest.Digest, error) {
+	defer resp.Body.Close()
+
+	body := limitReader(resp.Body, maxMetadataBytes)
+	content, err := io.ReadAll(body)
+	if err != nil {
+		return "", fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+	resp.Body = io.NopCloser(bytes.NewReader(content))
+
+	return digest.FromBytes(content), nil
+}
+
 // verifyContentDigest verifies "Docker-Content-Digest" header if present.
 // OCI distribution-spec states the Docker-Content-Digest header is optional.
 // Reference: https://github.com/opencontainers/distribution-spec/blob/v1.0.1/spec.md#legacy-docker-support-http-headers
 func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
+	digestStr := resp.Header.Get(dockerContentDigestHeader)
+
+	if len(digestStr) == 0 {
 		return nil
 	}
+
 	contentDigest, err := digest.Parse(digestStr)
 	if err != nil {
-		return fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+		return fmt.Errorf(
+			"%s %q: invalid response header: `%s: %s`",
+			resp.Request.Method, resp.Request.URL,
+			dockerContentDigestHeader, digestStr,
+		)
 	}
-	if contentDigest != expected {
-		return fmt.Errorf("%s: mismatch digest: %s", expected, contentDigest)
+
+	if len(expected) > 0 && contentDigest != expected {
```

Comment:
I'm curious that in which situation we will pass an empty `expected`?

Reply:
Removing that check, the following tests fail...
```
--- FAIL: Test_generateBlobDescriptorWithVariousDockerContentDigestHeaders (0.00s)
    repository_test.go:2576: [Blob.GET] 3. Client:Tag & Server:DigestWrongButSyntacticallyValid; expected no error for request, but got err: GET <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` vs expected ``
    repository_test.go:2576: [Blob.HEAD] 3. Client:Tag & Server:DigestWrongButSyntacticallyValid; expected no error for request, but got err: HEAD <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` vs expected ``
    repository_test.go:2576: [Blob.GET] 2. Client:Tag & Server:DigestValid; expected no error for request, but got err: GET <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:b526a4f2be963a2f9b0990c001255669eab8a254ab1a6e3f84f1820212ac7078` vs expected ``
    repository_test.go:2576: [Blob.HEAD] 2. Client:Tag & Server:DigestValid; expected no error for request, but got err: HEAD <nil>: invalid response; digest mismatch: `Docker-Content-Digest: sha256:b526a4f2be963a2f9b0990c001255669eab8a254ab1a6e3f84f1820212ac7078` vs expected ``
```

Going to have some food, then think about this a bit more.

Reply:
I think there's something wrong with the tests. Let me take a deeper look.

Reply:
I've checked the related code. The `refDigest` in `generateBlobDescriptor()` should never be empty. This is guaranteed by other checks before invoking `generateBlobDescriptor()`.
```go
func generateBlobDescriptor(resp *http.Response, refDigest digest.Digest)
```
Therefore, we should revise the related tests and remove the check of `len(expected) > 0`.

Reply:
Okay, the code has been reverted (no check for `len(expected)` just for the sake of testing), and instead, the test has been updated to account for this otherwise impossible scenario.